streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBWithTimestampStore.java

```diff
@@ -0,0 +1,679 @@
+                db.delete(noTimestampColumnFamily, wOptions, rawKey);
+                db.put(withTimestampColumnFamily, wOptions, rawKey, rawValueWithTimestamp);
+            } catch (final RocksDBException e) {
+                // String format is happening in wrapping stores. So formatted message is thrown from wrapping stores.
+                throw new ProcessorStateException("Error while putting key %s value %s into store " + name, e);
+            }
+        }
+    }
+
+    void write(final WriteBatch batch) throws RocksDBException {
+        db.write(wOptions, batch);
+    }
+
+    @Override
+    public void putAll(final List<KeyValue<Bytes, byte[]>> entries) {
+        try (final WriteBatch batch = new WriteBatch()) {
+            for (final KeyValue<Bytes, byte[]> entry : entries) {
+                Objects.requireNonNull(entry.key, "key cannot be null");
+                if (entry.value == null) {
+                    batch.delete(noTimestampColumnFamily, entry.key.get());
+                    batch.delete(withTimestampColumnFamily, entry.key.get());
+                } else {
+                    batch.delete(noTimestampColumnFamily, entry.key.get());
+                    batch.put(withTimestampColumnFamily, entry.key.get(), entry.value);
+                }
+            }
+            write(batch);
+        } catch (final RocksDBException e) {
+            throw new ProcessorStateException("Error while batch writing to store " + name, e);
+        }
+
+    }
+
+    @Override
+    public synchronized byte[] delete(final Bytes key) {
+        Objects.requireNonNull(key, "key cannot be null");
+        final byte[] valueWithTimestamp = get(key);
+        put(key, null);
+        return valueWithTimestamp;
+    }
+
+    @Override
+    public synchronized KeyValueIterator<Bytes, byte[]> range(final Bytes from,
+                                                              final Bytes to) {
+        Objects.requireNonNull(from, "from cannot be null");
+        Objects.requireNonNull(to, "to cannot be null");
+        validateStoreOpen();
+
+        // query rocksdb
+        final RocksDBRangeIterator rocksDBRangeIterator = new RocksDBRangeIterator(
+            name,
+            db.newIterator(withTimestampColumnFamily),
+            db.newIterator(noTimestampColumnFamily),
+            from,
+            to);
+        openIterators.add(rocksDBRangeIterator);
+
+        return rocksDBRangeIterator;
+    }
+
+    @Override
+    public synchronized KeyValueIterator<Bytes, byte[]> all() {
+        validateStoreOpen();
+        // query rocksdb
+        final RocksIterator innerIterWithTimestamp = db.newIterator(withTimestampColumnFamily);
+        innerIterWithTimestamp.seekToFirst();
+        final RocksIterator innerIterNoTimestamp = db.newIterator(noTimestampColumnFamily);
+        innerIterNoTimestamp.seekToFirst();
+        final RocksDbIterator rocksDbIterator = new RocksDbIterator(name, innerIterWithTimestamp, innerIterNoTimestamp);
```

Comment:
Will this work with callers? Will we inspect each value and insert an unknown timestamp for records from default column family?

EDIT: NM I didn't see the `RocksDbIterator` class below when writing this comment.